core/dataset.cpp

```diff
@@ -448,6 +448,10 @@ units::Unit DataConstProxy::unit() const {
   throw std::runtime_error("Data without values, unit is undefined.");
 }
 
+/// Return the name of the data proxy under which it is labeled inside the
+/// parent Dataset.
+std::string DataConstProxy::name() const { return m_name; }
```

Comment:
I agree that having a name is useful for certain purposes described in the PR, but it also break the logic for some existing behavior. For example, consider:

```python
d -= d['a']
```

Currently this will subtract `'a'` from *all* data items, since items are considered nameless. If we give it a name this behavior will suddenly be surprising and, I think, less intuitive.

Iteration of `Dataset` returns key/value pairs. Having the nam part of the value *in addition* to being the key is confusing.

Right now I do not know what the best way out is. Maybe just restrict the access to the name to friends, or rename `name()` to something non-obvious like `name_in_parent()`.

Reply:
I'm not sure i understand why the example you give would have a surprising behaviour. The name i added is only so that `d['a']` is aware that it is called `'a'` inside `d`. Are you saying that by writing `d -= d['a']`, the user might expect that we are removing the `'a'` variable from `d`, so that we are only left with e.g. variables `'b'` and `'c'`, instead of subtracting the values of `'a'` from the values of all data items? I don't really think this would be a problem, since the user does not need to be aware of any names. It is more for internal purposes.
All variables are still nameless inside `d`, a name is only given on-the-fly to `d['a']` as it is created.
Maybe `name_in_parent()` is a better option, as you suggest.

Reply:
> Are you saying that by writing `d -= d['a']`, the user might expect that we are removing the `'a'` variable from `d`, so that we are only left with e.g. variables `'b'` and `'c'`, instead of subtracting the values of `'a'` from the values of all data items?

No.

What I am saying is maybe best imagined via a table:
```
a | b | c
1 | 1 | 1 
1 | 1 | 1
```

Considering `d -= d['a']`:

- If `d['a']` has a name, i.e., is the column including its header, then the binary operation is a bit ambiguous (users might expect to subtract only from column `a`, just like subtracting two datasets from each other will match columns by name.
- If `d['a']` has no name, i.e., is the column without header, then it is (hopefully) clear that the operation will operate on all columns.

Reply:
> Maybe `name_in_parent()` is a better option, as you suggest.

#420 adds `DataArray`, which is also nameless. Ca we rethink the requirement for having a name?